### PR TITLE
[Randomness part 4] Update math/rand usage in crypto module and improve randomness tests

### PR DESCRIPTION
**Pull request by tarakby**
( part of this PR is a result of splitting https://github.com/onflow/flow-go/pull/4052 into several PRs)
  
  - prepare for Go1.20 update by removing deprecated `math/rand` functions `Seed` and `Read`.
  - improve randomness tests in (`crypto/random`, `utils/rand` and `fvm/environment`):
       - fix test bugs in `utils/rand` is extremely rare cases.
       - consolidate the statistic test tools in under crypto library - once the new crypto module is tagged, flow-go test would use the newly exported functions.
       - improve coverage of integer modulo-type random functions to cover small and large modulos.
       - improve the sample size to avoid bias of distribution and speed up tests when modulo is small.

### CODE REVIEW

File: crypto/random/rand_utils.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"gonum.org/v1/gonum/stat"
+)
+
+// BasicDistributionTest is a test function to run a basic statistic test on `randf` output.
```
Comment (tarakby) [inline crypto/random/rand_utils.go:12]: Randomness test tools consolidated here so that all `flow-go` tests can use them. 
Redundant code will be deleted once the new  crypto version is tagged.

File: fvm/environment/unsafe_random_generator_test.go [hunk 2]
```diff
@@ -16,30 +16,49 @@ import (
 	"github.com/onflow/flow-go/utils/unittest"
 )
 
+// TODO: these functions are copied from flow-go/crypto/rand
+// Once the new flow-go/crypto/ module version is tagged, flow-go would upgrade
+// to the new version and import these functions
+func BasicDistributionTest(t *testing.T, n uint64, classWidth uint64, randf func() (uint64, error)) {
```
Comment (tarakby) [inline fvm/environment/unsafe_random_generator_test.go:22]: redundant code till the new crypto tag is ready

File: utils/rand/rand_test.go [hunk 3]
```diff
@@ -9,90 +9,103 @@ import (
...
 )
 
-// Evaluate if the input distribution is close to uinform through a basic quick test.
-// The test computes the standard deviation and checks it is small enough compared
-// to the distribution mean.
-func evaluateDistributionUniformity(t *testing.T, distribution []float64) {
+// TODO: these functions are copied from flow-go/crypto/rand
+// Once the new flow-go/crypto/ module version is tagged, flow-go would upgrade
+// to the new version and import these functions
+func BasicDistributionTest(t *testing.T, n uint64, classWidth uint64, randf func() (uint64, error)) {
```
Comment (tarakby) [inline utils/rand/rand_test.go:19]: redundant code till the new crypto tag is ready

File: crypto/random/rand.go [hunk 4]
```diff
@@ -149,11 +149,7 @@ func (p *genericPRG) Shuffle(n int, swap func(i, j int)) error {
 	if n < 0 {
 		return fmt.Errorf("population size cannot be negative")
 	}
-	for i := n - 1; i > 0; i-- {
```
Comment (pattyshack) [inline crypto/random/rand.go:152]: the algorithms in utils/rand/rand.go is basically identical to this.  the only difference is the random source.  can you refactor the code to share implementation?

btw, Permutation / SubPermutation should also just call Shuffle/Sample

maybe undo changes to this file in this PR (since it's unrelated to test updates), and make the changes in a separate PR

**Review by pattyshack — APPROVED** [review]

File: crypto/bls12381_utils_test.go [hunk 5]
```diff
@@ -51,7 +51,7 @@ func TestPRGseeding(t *testing.T) {
 func BenchmarkScalarMultG1G2(b *testing.B) {
 	blsInstance.reInit()
 	seed := make([]byte, securityBits/8)
-	_, _ = rand.Read(seed)
+	_, _ = crand.Read(seed)
```
Comment (peterargue) [inline crypto/bls12381_utils_test.go:54]: can we check that error is nil?

File: crypto/bls12381_utils_test.go [hunk 6]
```diff
@@ -139,7 +139,7 @@ func TestSubgroupCheck(t *testing.T) {
 	blsInstance.reInit()
 	// seed Relic PRG
 	seed := make([]byte, securityBits/8)
-	_, _ = rand.Read(seed)
+	_, _ = crand.Read(seed)
```
Comment (peterargue) [inline crypto/bls12381_utils_test.go:142]: same here

File: crypto/bls_test.go [hunk 7]
```diff
@@ -784,7 +769,7 @@ func alterSignature(s Signature) {
 func BenchmarkBatchVerify(b *testing.B) {
 	// random message
 	input := make([]byte, 100)
-	_, _ = mrand.Read(input)
+	_, _ = crand.Read(input)
```
Comment (peterargue) [inline crypto/bls_test.go:772]: check for error?

File: crypto/bls_thresholdsign_test.go [hunk 8]
```diff
@@ -608,7 +609,7 @@ func testCentralizedStatelessAPI(t *testing.T) {
 func BenchmarkSimpleKeyGen(b *testing.B) {
 	n := 60
 	seed := make([]byte, SeedMinLenDKG)
-	_, _ = rand.Read(seed)
+	_, _ = crand.Read(seed)
```
Comment (peterargue) [inline crypto/bls_thresholdsign_test.go:612]: check for error?

File: crypto/ecdsa_test.go [hunk 9]
```diff
@@ -348,7 +353,7 @@ func TestEllipticUnmarshalSecp256k1(t *testing.T) {
 func BenchmarkECDSADecode(b *testing.B) {
 	// random message
 	seed := make([]byte, 50)
-	_, _ = rand.Read(seed)
+	_, _ = crand.Read(seed)
```
Comment (peterargue) [inline crypto/ecdsa_test.go:356]: check for error?

**Review by peterargue — APPROVED** [review]
looks good. a few nits

File: crypto/hash/hash_test.go [hunk 10]
```diff
@@ -266,7 +257,7 @@ func TestKeccak(t *testing.T) {
 func BenchmarkComputeHash(b *testing.B) {
 
 	m := make([]byte, 32)
-	rand.Read(m)
+	_, _ = rand.Read(m)
```
[the remark this answers is not in the document]
Comment (tarakby) [reply crypto/hash/hash_test.go:260]: In these bench functions, if randomness fail, we would measure the hash of all zeros instead of a random input. 
I'll add the check to avoid confusion.

File: crypto/bls12381_utils_test.go (hunk 5 above)
Comment (tarakby) [reply crypto/bls12381_utils_test.go:54]: In these bench functions, if randomness fail, we would measure the timing on the same inputs instead of random inputs. 
I'll add the check to avoid confusion.

File: crypto/bls_test.go (hunk 7 above)
Comment (tarakby) [reply crypto/bls_test.go:772]: Same here, but I'll add the check

File: crypto/random/rand.go (hunk 4 above)
Comment (tarakby) [reply crypto/random/rand.go:152]: > Permutation / SubPermutation should also just call Shuffle/Sample

This is true, though it is slightly less efficient: in order to use `Shuffle`, `Permutation` needs to initialize an array firts, and then shuffle it. The current implementation of `Permutation` initializes and permutes at the same time (it's an optimized [variant](https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle#The_%22inside-out%22_algorithm) of Fisher-Yates).

> can you refactor the code to share implementation?

You are right that the core idea is the same. I did explore the idea of combining both implementations but I didn't find an elegant way to do that, given there are minor differences between the packages (returned errors, thread safety, integer types). For now I only combined the test tools, but we can discuss ways to combine the rest.
